crates/task-impls/src/quorum_vote.rs

```diff
@@ -0,0 +1,383 @@
+            }),
+        );
+    }
+
+    /// Update the latest voted view number.
+    #[instrument(skip_all, fields(id = self.id, latest_voted_view = *self.latest_voted_view), name = "Quorum vote update latest voted view", level = "error")]
+    async fn update_latest_voted_view(&mut self, new_view: TYPES::Time) -> bool {
+        if *self.latest_voted_view < *new_view {
+            debug!(
+                "Updating next vote view from {} to {} in the quorum vote task",
+                *self.latest_voted_view, *new_view
+            );
+
+            // Cancel the old dependency tasks.
+            for view in (*self.latest_voted_view + 1)..=(*new_view) {
+                if let Some(dependency) = self.vote_dependencies.remove(&TYPES::Time::new(view)) {
+                    cancel_task(dependency).await;
+                }
+            }
+
+            self.latest_voted_view = new_view;
+
+            return true;
+        }
+        false
+    }
+
+    /// Handles a consensus event received on the event stream
+    #[instrument(skip_all, fields(id = self.id, latest_voted_view = *self.latest_voted_view), name = "Quorum vote handle", level = "error")]
+    pub async fn handle(
+        &mut self,
+        event: Arc<HotShotEvent<TYPES>>,
+        event_receiver: Receiver<Arc<HotShotEvent<TYPES>>>,
+        event_sender: Sender<Arc<HotShotEvent<TYPES>>>,
+    ) {
+        match event.as_ref() {
+            HotShotEvent::QuorumProposalRecv(proposal, _sender) => {
+                let view = proposal.data.view_number;
+                if view <= self.latest_voted_view {
+                    return;
+                }
+                debug!("Received Quorum Proposal for view {}", *view);
+
+                // stop polling for the received proposal
+                self.quorum_network
+                    .inject_consensus_info(ConsensusIntentEvent::CancelPollForProposal(*view))
+                    .await;
+                broadcast_event(Arc::new(HotShotEvent::ViewChange(view + 1)), &event_sender).await;
+                if !validate_quorum_proposal(proposal.clone(), event_sender.clone()) {
+                    return;
+                }
+                broadcast_event(
+                    Arc::new(HotShotEvent::QuorumProposalValidated(proposal.data.clone())),
+                    &event_sender.clone(),
+                )
+                .await;
+                self.create_dependency_task_if_new(view, event_receiver, event_sender);
+            }
+            HotShotEvent::DACRecv(cert) => {
+                debug!("DAC Received for view {}!", *cert.view_number);
+                let view = cert.view_number;
+
+                self.quorum_network
+                    .inject_consensus_info(ConsensusIntentEvent::CancelPollForDAC(*view))
+                    .await;
+
+                self.committee_network
+                    .inject_consensus_info(ConsensusIntentEvent::CancelPollForVotes(*view))
+                    .await;
```

Comment:
It'll be nice when half of this function disappears once we can get rid of the Polling.